File: Java/src/main/java/edu/bu/benchmarks/RandomRead.java
package edu.bu.benchmarks;

import edu.bu.filehandler.FileHandler;
import edu.bu.filehandler.LogFileHandler;
import edu.bu.util.RootData;
import org.apache.log4j.PropertyConfigurator;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.stream.Stream;

public class RandomRead {

    public static void main(String[] args) throws IOException {

        Properties prop = new Properties();

        PropertyConfigurator.configure(prop);
        prop.setProperty("log4j.appender.WORKLOG.File", "log4j.properties");

        PropertyConfigurator.configure("log4j.properties");

        //Set serialized file path

        String inFile = args[0];

        int serializationType = Integer.parseInt(args[1]);

        long cunt_to_read = Integer.parseInt(args[2]);

        int round=Integer.parseInt(args[3]);

        boolean taskset=Boolean.parseBoolean(args[5]);

        List<Long>  objectIndexes=new ArrayList<>();
        // read random data from file
        try (Stream<String> stream = Files.lines(Paths.get(args[4]))) {
            stream.forEach(
                    e -> {
                        objectIndexes.add(Long.parseLong(e));
                    }
            );
        }

        String logFileName="bin/benchmark/readobjects/result_java_readobjects_" + cunt_to_read + "_" + round + ".txt";

        FileHandler fileHandler = new FileHandler(inFile, serializationType);
        fileHandler.prepareToRead();

        // Start Calculation time:
        long tmpTime = System.nanoTime();

        List<RootData>  objectList=new ArrayList<>();
        for (long j = 0; j < cunt_to_read; ++j) {
           objectList.add(fileHandler.getObjectsFromFile(objectIndexes.get((int) j), 1).get(0));
        }

        // Time Calculation
        double elapsedSeconds= (System.nanoTime() - tmpTime) / 1000000000.0;

        LogFileHandler logFileHandler=new LogFileHandler(logFileName);
        logFileHandler.addLog(true,serializationType,false,"TweetStatus",fileHandler.getIoTime(),elapsedSeconds,taskset);

    }
}

File: Rust/src/benchmarks/RandomRead.rs
use std::io;
use crate::filehandler::FileHandler::FileHandler;
use std::time:: Instant;
use crate::tweetStructs::TweetStatus::TweetStatus;
use crate::filehandler::LogFileHandler::LogFileHandler;
use std::fs::File;
use std::io::{BufReader, BufRead};
use bytes::{BytesMut, Buf};

pub fn RandomRead(args:Vec<String> ) -> io::Result<()>{

    let inFile:String = args[2].to_string();

    let serializationType:i32 = args[3].parse().unwrap();

    //Set output file path
    let random_list_path:String = args[4].to_string();

    // set number of tweets:
    let cunt_to_read:u64=args[5].parse().unwrap();

    let  round:i32=args[6].parse().unwrap();

    let  taskset:bool= args[7].parse().unwrap();

    let mut objectIndexes:Vec<u64>=vec![];

    // read random data from file
    let f = File::open(&random_list_path)?;
    let f = BufReader::new(f);
    for line in f.lines() {
        objectIndexes.push(line.unwrap().parse().unwrap());
    }


    let log_file_name:String = format!("{}{}{}{}{}", "bin/benchmark/readobjects/result_rust_readobjects_", cunt_to_read, "_", round, ".txt");
    let logfile=LogFileHandler::new(String::from(log_file_name));

    let mut file_handler = FileHandler::new(inFile, serializationType);
    file_handler.prepareToRead();

    let start = Instant::now();
    let mut objectList:Vec<TweetStatus>=vec![];

    for u in objectIndexes {
        file_handler.getObjectsFromFile(u,1, &mut objectList);
    }
    let elapsedSeconds = start.elapsed();
    logfile.add_log(true, serializationType, false, String::from("TweetStatus"), file_handler.getio_time(), elapsedSeconds, taskset);

    Ok(())
}